core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala

```diff
@@ -0,0 +1,292 @@
+ */
+class PartitionLockTest extends Logging {
+
+  val numReplicaFetchers = 2
+  val numProducers = 3
+  val numRecordsPerProducer = 5
+
+  val mockTime = new MockTime() // Used for check to shrink ISR
+  val tmpDir = TestUtils.tempDir()
+  val logDir = TestUtils.randomPartitionLogDir(tmpDir)
+  val executorService = Executors.newFixedThreadPool(numReplicaFetchers + numProducers + 1)
+  val appendSemaphore = new Semaphore(0)
+  val shrinkIsrSemaphore = new Semaphore(0)
+
+  var logManager: LogManager = _
+  var partition: Partition = _
+
+
+  @Before
+  def setUp(): Unit = {
+    val logConfig = new LogConfig(new Properties)
+    logManager = TestUtils.createLogManager(Seq(logDir), logConfig, CleanerConfig(enableCleaner = false), mockTime)
+    partition = setupPartitionWithMocks(logManager, logConfig)
+  }
+
+  @After
+  def tearDown(): Unit = {
+    executorService.shutdownNow()
+    logManager.liveLogDirs.foreach(Utils.delete)
+    Utils.delete(tmpDir)
+  }
+
+  /**
+   * Verifies that delays in appending to leader while processing produce requests has no impact on timing
+   * of update of log read result when processing replica fetch request if no ISR update is required.
+   */
+  @Test
+  def testNoLockContentionWithoutIsrUpdate(): Unit = {
+    concurrentProduceFetch(numProducers, numReplicaFetchers, numRecordsPerProducer, None)
+  }
+
+  /**
+   * Verifies that delays in appending to leader while processing produce requests has no impact on timing
+   * of update of log read result when processing replica fetch request even if a scheduler thread is checking
+   * for ISR shrink conditions if no ISR update is required.
+   */
+  @Test
+  def testAppendReplicaFetchWithSchedulerCheckForShrinkIsr(): Unit = {
+    val active = new AtomicBoolean(true)
+
+    val future = scheduleShrinkIsr(active, mockTimeSleepMs = 0)
+    concurrentProduceFetch(numProducers, numReplicaFetchers, numRecordsPerProducer, None)
+    active.set(false)
+    future.get(15, TimeUnit.SECONDS)
+  }
+
+  /**
+   * Verifies concurrent produce and replica fetch log read result update with ISR updates. This
+   * can result in delays in processing produce and replica fetch requets since write lock is obtained,
+   * but it should complete without any failures.
+   */
+  @Test
+  def testAppendReplicaFetchWithUpdateIsr(): Unit = {
+    val active = new AtomicBoolean(true)
+
+    val future = scheduleShrinkIsr(active, mockTimeSleepMs = 10000)
+    TestUtils.waitUntilTrue(() => shrinkIsrSemaphore.hasQueuedThreads, "shrinkIsr not invoked")
+    concurrentProduceFetch(numProducers, numReplicaFetchers, numRecordsPerProducer, Some(shrinkIsrSemaphore))
+    active.set(false)
+    future.get(15, TimeUnit.SECONDS)
+  }
+
+  private def concurrentProduceFetch(numProducers: Int,
+                                     numReplicaFetchers: Int,
+                                     numRecords: Int,
+                                     contentionSemaphore: Option[Semaphore]): Unit = {
```

Comment:
nit: would be nice to use a consistent mechanism to propagate semaphores to threads. Also, should this be `shrinkIsrSemaphore`?